Fast forward scheduler

ducky64 opened the pull request:
Depends on #225, JDOM test utilities, should not merge until after that goes in.

First pass of a fast-forward scheduler, an alternative to the wall clock scheduler that simulates events as fast as possible, for use in unit testing. The button tests from #225 run significantly faster with this PR. The simulation only advances through calls to `runToPromise`, which runs simulation cycles until the argument promise is resolved. On each simulated callback (`scheduler.setInterval` or `setTimeout`), the scheduler yields to the event queue to allow async tasks to run (but this probably isn't bulletproof, since any `setTimeout`s called from those will not run until after the scheduler re-runs).

Pulls in a priority queue dependency, which is used to find the next event in an efficient way.

Open questions (see comments thread for resolution):
- Is Promises as an API for controlling time advance the right API? It's consistent with how async / await are used in Typescript, but the resolve chain for a promise is not run immediately (instead, scheduled on the event loop). This means that we have to yield in the scheduler and hope everything finishes before the scheduler goes around the event loop again
- As a hack, the functions in tester.ts how call `runToPromise`, which means they will only work with buses created for testing. Where should `runToPromise` go, and is there a good way to support both the fast-forward scheduler and wall-clock scheduler (and possibly other schedulers we come up with down the road) in the test API? Some other options:
  - Check the scheduler type in the test API, special-case where we have a fast-forward / nonrealtime scheduler (can be subclass / instanceof based?). May not support other scheduler types that pop up in the future.
  - Add runToPromise (or similar) into the base Scheduler interface. Does not seem like the right solution.

TODO:
- [x] Needs test for clearTimeout, clearInterval
- [x] Needs test for clearInterval after an interval has elapsed
- [x] Run prettier

Resolves microsoft/jacdac#506

pelikhan commented on src/jdom/scheduler.ts:
```diff
@@ -1,3 +1,6 @@
+import { assert } from "./jacdac-jdom"
```
since this scheduler is bringing a new dependency (heap), we should move it out of the jdom library and in the test folder. It's generally not used outside of unit testing.

ducky64 replied:
Done (I think?).

pelikhan commented on src/jdom/scheduler.ts:
```diff
@@ -1,3 +1,6 @@
+import { assert } from "./jacdac-jdom"
```
don't import from "./jacdac-dom" here, instead import from the utils file.

pelikhan commented on src/jdom/scheduler.ts:
```diff
@@ -55,3 +58,121 @@ export class WallClockScheduler implements Scheduler {
+        delay: number,
+        ...args: any[]
+    ): any {
+        const eventQueueElt = {
```
is this an EventRecord ? 
```
const eventQueueElt : EventRecord = {
```

pelikhan commented on src/jdom/scheduler.ts:
```diff
@@ -55,3 +58,121 @@ export class WallClockScheduler implements Scheduler {
+        delay: number,
+        ...args: any[]
+    ): any {
+        const eventQueueElt = {
```
strongly type eventQueueElt

ducky64 commented:
Should be ready for a merge review now - not completely happy with everything but a reasonable good-enough point.

- Moved FastForwardScheduler out of JDOM and into the test directory, and moved the heap dependency to devDependencies.
- I chose to have nextEventFrom and friends detect if it's a FastForwardScheduler and act accordingly. Not a perfect solution, but since FastForwardScheduler is now in the tests directory, it seems more reasonable.
  - Architecturally, this breaks nextEventFrom (wrapper that detects FastForwardScheduler) from nextEventFromInternal (which contains the core logic only, assuming the scheduler / wall time is running)
  - A more architecturally cleaner but heavyweight solution is to have withBus return a special BusTester object which encapsulates the fast-forward scheduler and exposes test functions that behaves accordingly. A separate BusTester class could be defined to work with the wall-time scheduler.
- Added unit tests for setInterval, setTimeout, clearInterval, clearTimeout
- Refactors nextEventFrom to be async instead of using promises. Should be a lot cleaner now.

pelikhan commented:
How does with test bus decide to use FF scheduler vs wall time?

pelikhan reviewed (approved):
Good iteration!

ducky64 commented:
`withTestBus` is hardcoded to use the FF scheduler. It could probably take an options object in the future if needed, but I doubt it would make sense to use a wallclock scheduler in a test, barring potential future hardware-in-the-loop tests.
